File: usage-daemon/src/cookiejar.js
// Read a provider's session cookie out of Firefox, on request.
//
// The problem: ollama, mistral and opencode-go authenticate with a browser
// session cookie. Those expire. Every expiry meant noticing the provider had
// gone `auth_expired`, opening devtools, copying the Cookie header by hand,
// and pasting it into the web UI. Firefox already holds a live copy of exactly
// that cookie — so on the rare occasion it goes stale, ask Firefox instead of
// asking the human to transcribe it.
//
// ON DEMAND, NEVER ON A TIMER. An earlier cut re-read Firefox before every
// poll. That was wrong twice over: it gave a long-running daemon a standing
// capability to harvest browser cookies unattended, and it copied a ~2 MB
// database every few minutes to re-learn something that only changes when a
// session expires. The daemon now touches the profile only when a human
// explicitly asks — see Runner.refreshCookieFromFirefox().
//
// Mechanics:
//   - `node:sqlite` (Node 22+) reads cookies.sqlite. Already used by the
//     opencode-go plugin, so no new dependency.
//   - Firefox stores cookie values in PLAINTEXT — no OS-keyring decrypt. (That
//     complexity is Chromium-specific; confirmed on this host 2026-08-16.)
//   - The database is copied to a temp file before opening, because Firefox
//     keeps it in WAL mode while running and a read-only open can fail when
//     SQLite wants to recover the WAL. Copying also guarantees this can never
//     disturb Firefox's own state. The profile is never written to.
//
// Cookie VALUES are secrets: returned to the caller for the outbound request,
// never logged. Log lines here carry counts, hostnames and expiry only.

import { DatabaseSync } from 'node:sqlite';
import { copyFileSync, existsSync, mkdtempSync, readdirSync, rmSync, statSync } from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { log } from './log.js';

function firefoxRoot() {
  return process.env.USAGE_FIREFOX_DIR || path.join(os.homedir(), '.mozilla', 'firefox');
}

// Pick the profile actually in use: prefer `default-release`, then whichever
// cookies.sqlite was modified most recently. A stale `default-esr` profile
// sitting alongside the live one is the common case here, and would otherwise
// hand back cookies that expired months ago.
export function findProfileDb(root = firefoxRoot()) {
  if (!existsSync(root)) return null;
  const candidates = [];
  for (const entry of readdirSync(root)) {
    const db = path.join(root, entry, 'cookies.sqlite');
    if (!existsSync(db)) continue;
    candidates.push({ profile: entry, db, mtime: statSync(db).mtimeMs });
  }
  if (!candidates.length) return null;
  candidates.sort((a, b) => {
    const aDef = a.profile.includes('default-release') ? 1 : 0;
    const bDef = b.profile.includes('default-release') ? 1 : 0;
    if (aDef !== bDef) return bDef - aDef;
    return b.mtime - a.mtime;
  });
  return candidates[0];
}

// Firefox stores host as either `example.com` or `.example.com` (the latter
// meaning "and every subdomain"). Match the domain plus any subdomain, so
// `mistral.ai` picks up `.mistral.ai`, `console.` and `admin.` together — the
// mistral plugin calls two of those hosts, and the pasted cookie that worked
// was always a merge across all of them.
export function hostMatches(cookieHost, domain) {
  const h = String(cookieHost ?? '').replace(/^\./, '').toLowerCase();
  const d = String(domain ?? '').toLowerCase();
  if (!h || !d) return false;
  return h === d || h.endsWith(`.${d}`);
}

// Serialize to what a browser would put in the Cookie header. Later duplicates
// win: rows arrive shortest-host-first, so a subdomain's cookie overrides the
// bare domain's cookie of the same name — the more specific one is what the
// site set for the page being called.
export function toCookieHeader(rows) {
  const seen = new Map();
  for (const r of rows) {
    if (!r?.name) continue;
    seen.set(r.name, r.value ?? '');
  }
  return [...seen.entries()].map(([k, v]) => `${k}=${v}`).join('; ');
}

// Normalize moz_cookies.expiry to epoch SECONDS.
//
// The unit is not stable across Firefox versions: this profile (Firefox 153)
// stores milliseconds — 13 digits, e.g. 1787030245429 — while older profiles
// store seconds. Getting this wrong is not cosmetic: comparing a millisecond
// stamp against a seconds-based `now` makes EVERY cookie look unexpired, which
// silently defeats the expired-cookie filter and hands out a dead header that
// then fails authentication for no visible reason. Observed here 2026-08-18.
//
// Ranges (all far apart, so the guess is unambiguous for any real cookie):
//   < 1e11  seconds       (1e11s is the year 5138; real values ~1.8e9)
//   < 1e14  milliseconds  (~1.787e12 today)
//   else    microseconds  (Firefox uses these for creationTime/lastAccessed)
export function toEpochSeconds(expiry) {
  if (typeof expiry !== 'number' || !Number.isFinite(expiry) || expiry <= 0) return null;
  if (expiry < 1e11) return expiry;
  if (expiry < 1e14) return expiry / 1e3;
  return expiry / 1e6;
}

// Soonest real expiry across the cookies making up the header — i.e. when this
// header starts to degrade. A session cookie (expiry 0) has no date to report.
export function soonestExpiry(rows) {
  const stamps = rows.map((r) => toEpochSeconds(r.expiry)).filter((e) => e != null);
  if (!stamps.length) return null;
  return new Date(Math.min(...stamps) * 1000).toISOString();
}

// Copy the DB (plus its WAL sidecars) somewhere private, read it, clean up.
function readRows(dbPath, domain) {
  const tmpDir = mkdtempSync(path.join(os.tmpdir(), 'usage-daemon-cookies-'));
  const copy = path.join(tmpDir, 'cookies.sqlite');
  try {
    copyFileSync(dbPath, copy);
    for (const suffix of ['-wal', '-shm']) {
      if (existsSync(dbPath + suffix)) copyFileSync(dbPath + suffix, copy + suffix);
    }
    const db = new DatabaseSync(copy, { readOnly: true });
    try {
      // Shortest host first, so bare-domain rows land before the more specific
      // subdomain rows that should override them in toCookieHeader().
      return db
        .prepare(
          `SELECT host, name, value, expiry FROM moz_cookies
            WHERE host LIKE ? OR host LIKE ?
            ORDER BY length(host) ASC`,
        )
        .all(`%${domain}`, `%.${domain}`);
    } finally {
      db.close();
    }
  } finally {
    rmSync(tmpDir, { recursive: true, force: true });
  }
}

// Build the Cookie header for `domain` from Firefox's cookie store.
// Returns { header, expiresAt, count }; `header` is '' when there is nothing
// usable, and the caller reports that to the user rather than storing it.
export function cookieHeaderFor(domain) {
  const empty = { header: '', expiresAt: null, count: 0 };

  const profile = findProfileDb();
  if (!profile) {
    log.warn('firefox cookie source: no profile found', { root: firefoxRoot(), domain });
    return empty;
  }

  let rows;
  try {
    rows = readRows(profile.db, domain);
  } catch (err) {
    log.warn('firefox cookie source: could not read cookie db', {
      domain, profile: profile.profile, err,
    });
    return empty;
  }

  const now = Date.now() / 1000;
  const matching = rows.filter((r) => hostMatches(r.host, domain));
  // A null normalized expiry means a session cookie: Firefox holds it for the
  // life of the browser session and it is perfectly valid to send.
  const live = matching.filter((r) => {
    const exp = toEpochSeconds(r.expiry);
    return exp == null || exp > now;
  });
  const expired = matching.filter((r) => !live.includes(r));
  const header = toCookieHeader(live);

  if (!header) {
    log.warn('firefox cookie source: no live cookies for domain', {
      domain,
      profile: profile.profile,
      matched: matching.length,
      expired: expired.length,
      hint: `log in to ${domain} in Firefox`,
    });
    return empty;
  }

  // Names, never values — enough to see exactly what was handed to the
  // provider (and what was dropped as stale) when a refresh misbehaves.
  log.info('firefox cookie source: built header', {
    domain,
    profile: profile.profile,
    hosts: [...new Set(live.map((r) => r.host))].join(','),
    cookies: live.map((r) => r.name).join(','),
    dropped_expired: expired.length ? expired.map((r) => r.name).join(',') : undefined,
    expires_at: soonestExpiry(live) ?? 'session-only',
  });

  return {
    header,
    expiresAt: soonestExpiry(live),
    count: live.length,
    profile: profile.profile,
    names: live.map((r) => r.name),
  };
}

File: usage-daemon-v3/usage_daemon/cookiejar.py
"""Read a provider's session cookie out of Firefox, on request (port of
src/cookiejar.js).

ON DEMAND, NEVER ON A TIMER. The daemon touches the Firefox profile only when a
human explicitly asks (Runner.refresh_cookie_from_firefox), never on a poll.

Mechanics:
  - Firefox stores cookie values in PLAINTEXT (no OS-keyring decrypt).
  - The DB is copied to a temp dir before opening: Firefox keeps it in WAL mode
    while running and a read-only open can fail when SQLite wants to recover the
    WAL. Copying also guarantees we never disturb Firefox's state.

Cookie VALUES are secrets: returned to the caller for the outbound request,
never logged. Log lines carry counts/hostnames/expiry only.
"""

from __future__ import annotations

import math
import os
import shutil
import sqlite3
import tempfile
import time

from .log import log


def firefox_root() -> str:
    return os.environ.get("USAGE_FIREFOX_DIR") or os.path.join(
        os.path.expanduser("~"), ".mozilla", "firefox"
    )


def find_profile_db(root: str | None = None) -> dict | None:
    """Pick the live profile: prefer default-release, else most-recently-modified."""
    root = root or firefox_root()
    if not os.path.isdir(root):
        return None
    candidates = []
    for entry in os.listdir(root):
        db = os.path.join(root, entry, "cookies.sqlite")
        if not os.path.isfile(db):
            continue
        try:
            mtime = os.path.getmtime(db)
        except OSError:
            mtime = 0
        candidates.append({"profile": entry, "db": db, "mtime": mtime})
    if not candidates:
        return None
    candidates.sort(key=lambda c: (c["profile"].count("default-release") > 0, c["mtime"]), reverse=True)
    return candidates[0]


def host_matches(cookie_host, domain: str) -> bool:
    h = (cookie_host or "").lstrip(".").lower()
    d = (domain or "").lower()
    if not h or not d:
        return False
    return h == d or h.endswith("." + d)


def to_epoch_seconds(expiry):
    """Normalize moz_cookies.expiry (units vary by Firefox version) to seconds.

    Fail-soft like the JS toEpochSeconds: junk (null, strings, NaN, Infinity,
    non-positive) has no expiry — NaN/Inf must not slip through the numeric
    branches (NaN <= 0 is False in Python) and poison expiry math.
    """
    if isinstance(expiry, bool) or not isinstance(expiry, (int, float)):
        return None
    if not math.isfinite(expiry) or expiry <= 0:
        return None
    if expiry < 1e11:
        return expiry  # seconds
    if expiry < 1e14:
        return expiry / 1e3  # milliseconds
    return expiry / 1e6  # microseconds


def to_cookie_header(rows) -> str:
    seen: dict[str, str] = {}
    for r in rows:
        if not r.get("name"):
            continue
        seen[r["name"]] = r.get("value") or ""
    return "; ".join(f"{k}={v}" for k, v in seen.items())


def soonest_expiry(rows) -> str | None:
    stamps = [to_epoch_seconds(r.get("expiry")) for r in rows]
    stamps = [s for s in stamps if s is not None]
    if not stamps:
        return None
    tm = min(stamps)
    from datetime import datetime, timezone

    return datetime.fromtimestamp(tm, timezone.utc).isoformat()


def _read_rows(db_path: str, domain: str) -> list[dict]:
    tmp_dir = tempfile.mkdtemp(prefix="usage-daemon-cookies-")
    copy = os.path.join(tmp_dir, "cookies.sqlite")
    rows = []
    try:
        shutil.copy2(db_path, copy)
        for suffix in ("-wal", "-shm"):
            if os.path.exists(db_path + suffix):
                try:
                    shutil.copy2(db_path + suffix, copy + suffix)
                except OSError:
                    pass
        conn = sqlite3.connect(f"file:{copy}?mode=ro", uri=True, check_same_thread=False)
        conn.row_factory = sqlite3.Row
        try:
            # Shortest host first so bare-domain rows land before subdomain rows
            # that should override them in to_cookie_header().
            cur = conn.execute(
                "SELECT host, name, value, expiry FROM moz_cookies "
                "WHERE host LIKE ? OR host LIKE ? "
                "ORDER BY length(host) ASC",
                (f"%{domain}", f"%.{domain}"),
            )
            for r in cur.fetchall():
                rows.append(dict(r))
        finally:
            conn.close()
    finally:
        shutil.rmtree(tmp_dir, ignore_errors=True)
    return rows


def cookie_header_for(domain: str) -> dict:
    """Build the Cookie header for `domain` from Firefox's store."""
    empty = {"header": "", "expiresAt": None, "count": 0}
    profile = find_profile_db()
    if profile is None:
        log.warn("firefox cookie source: no profile found", {"root": firefox_root(), "domain": domain})
        return empty
    try:
        rows = _read_rows(profile["db"], domain)
    except Exception as err:
        log.warn("firefox cookie source: could not read cookie db", {
            "domain": domain, "profile": profile["profile"], "err": err,
        })
        return empty

    now = time.time()
    matching = [r for r in rows if host_matches(r.get("host"), domain)]
    live = []
    for r in matching:
        exp = to_epoch_seconds(r.get("expiry"))
        if exp is None or exp > now:
            live.append(r)
    expired = [r for r in matching if r not in live]
    header = to_cookie_header(live)

    if not header:
        log.warn("firefox cookie source: no live cookies for domain", {
            "domain": domain, "profile": profile["profile"],
            "matched": len(matching), "expired": len(expired),
            "hint": f"log in to {domain} in Firefox",
        })
        return empty

    new_names = list({r.get("host") for r in live})
    log.info("firefox cookie source: built header", {
        "domain": domain,
        "profile": profile["profile"],
        "hosts": ",".join(sorted(new_names)),
        "cookies": ",".join(r.get("name", "") for r in live),
        "dropped_expired": ",".join(r.get("name", "") for r in expired) or None,
        "expires_at": soonest_expiry(live) or "session-only",
    })
    return {
        "header": header,
        "expiresAt": soonest_expiry(live),
        "count": len(live),
        "profile": profile["profile"],
        "names": [r.get("name", "") for r in live],
    }